File: service/app/schemas/rbac.py
"""
RBAC（基于角色的访问控制）相关Schema
"""
from pydantic import BaseModel, Field
from typing import List, Optional, Literal, TYPE_CHECKING
from datetime import datetime

if TYPE_CHECKING:
    from app.schemas.rbac import PermissionResponse


class PermissionBase(BaseModel):
    """权限基础模型"""
    name: str = Field(..., min_length=1, max_length=100, description="权限名称")
    code: str = Field(..., min_length=1, max_length=100, description="权限代码（唯一标识）")
    resource: str = Field(..., min_length=1, max_length=50, description="资源（如：tenant, prompts, rbac）")
    action: str = Field(..., min_length=1, max_length=50, description="操作（如：create, list, menu_list）")
    type: Literal["menu", "api", "button"] = Field("api", description="权限类型：menu=菜单, api=接口, button=按钮")
    type_name: Optional[str] = Field(None, max_length=50, description="类型名称，用于展示")
    group_name: Optional[str] = Field(None, max_length=50, description="分组名称，用于前端 groupBy")
    is_system_admin_only: bool = Field(False, description="是否仅系统管理员可见")
    description: Optional[str] = Field(None, max_length=500, description="权限描述")
    parent_id: Optional[str] = Field(None, description="父权限ID（用于菜单层级）")
    sort_order: int = Field(0, description="排序顺序（用于菜单排序）")


class PermissionCreate(PermissionBase):
    """创建权限请求模型"""
    is_active: bool = Field(True, description="是否激活")
    parent_id: Optional[str] = Field(None, description="父权限ID（用于菜单层级）")
    sort_order: int = Field(0, description="排序顺序（用于菜单排序）")


class PermissionUpdate(BaseModel):
    """更新权限请求模型"""
    name: Optional[str] = Field(None, min_length=1, max_length=100, description="权限名称")
    type: Optional[Literal["menu", "api", "button"]] = Field(None, description="权限类型")
    type_name: Optional[str] = Field(None, max_length=50, description="类型名称")
    group_name: Optional[str] = Field(None, max_length=50, description="分组名称")
    is_system_admin_only: Optional[bool] = Field(None, description="是否仅系统管理员可见")
    description: Optional[str] = Field(None, max_length=500, description="权限描述")
    parent_id: Optional[str] = Field(None, description="父权限ID（用于菜单层级）")
    sort_order: Optional[int] = Field(None, description="排序顺序（用于菜单排序）")
    is_active: Optional[bool] = Field(None, description="是否激活")


class PermissionResponse(PermissionBase):
    """权限响应模型"""
    id: str
    type: str = "api"
    type_name: Optional[str] = None
    group_name: Optional[str] = None
    is_system_admin_only: bool = False
    parent_id: Optional[str] = None
    sort_order: int = 0
    is_active: bool
    created_at: datetime
    updated_at: datetime
    children: Optional[List["PermissionResponse"]] = Field(default_factory=list, description="子权限列表（用于菜单树）")

    class Config:
        from_attributes = True


class MenuTreeNode(BaseModel):
    """菜单树节点模型（专门用于菜单树 API）"""
    id: str
    name: str
    code: str
    resource: str
    action: str
    type: str
    description: Optional[str] = None
    parent_id: Optional[str] = Field(None, description="父权限ID（None 表示根节点）")
    sort_order: int = Field(0, description="排序顺序")
    is_active: bool
    children: List["MenuTreeNode"] = Field(default_factory=list, description="子菜单列表")
    
    class Config:
        from_attributes = True
        populate_by_name = True  # 允许使用别名
        json_encoders = {
            # 确保正确序列化
        }


class RoleBase(BaseModel):
    """角色基础模型"""
    name: str = Field(..., min_length=1, max_length=100, description="角色名称")
    code: str = Field(..., min_length=1, max_length=100, description="角色代码（唯一标识）")
    description: Optional[str] = Field(None, max_length=500, description="角色描述")


class RoleCreate(RoleBase):
    """创建角色请求模型"""
    is_active: bool = Field(True, description="是否激活")
    permission_ids: Optional[List[str]] = Field(default_factory=list, description="权限ID列表")


class RoleUpdate(BaseModel):
    """更新角色请求模型"""
    name: Optional[str] = Field(None, min_length=1, max_length=100, description="角色名称")
    description: Optional[str] = Field(None, max_length=500, description="角色描述")
    is_active: Optional[bool] = Field(None, description="是否激活")
    permission_ids: Optional[List[str]] = Field(None, description="权限ID列表")


class RoleResponse(RoleBase):
    """角色响应模型"""
    id: str
    team_code: Optional[str] = Field(None, description="团队代码（None表示全局角色）")
    is_active: bool
    permissions: List[PermissionResponse] = Field(default_factory=list, description="权限列表")
    created_at: datetime
    updated_at: datetime
    
    class Config:
        from_attributes = True


class UserRoleAssign(BaseModel):
    """用户角色分配请求模型"""
    user_id: str = Field(..., description="用户ID")
    role_ids: List[str] = Field(..., min_items=0, description="角色ID列表")


class RolePermissionAssign(BaseModel):
    """角色权限分配请求模型"""
    role_id: str = Field(..., description="角色ID")
    permission_ids: List[str] = Field(..., min_items=0, description="权限ID列表")

File: app/lib/api/rbac.ts
/**
 * RBAC权限管理相关 API
 */

import { apiRequest, ApiError } from "./config";

/**
 * 权限类型：menu=菜单权限（控制侧栏与路由），api=接口权限
 */
export type PermissionType = "menu" | "api";

/**
 * 权限信息
 */
export interface Permission {
  id: string;
  name: string;
  code: string;
  resource: string;
  action: string;
  type?: PermissionType;
  description?: string;
  parent_id?: string;
  sort_order?: number;
  is_active: boolean;
  created_at: string;
  updated_at: string;
  children?: Permission[];
}

/**
 * 创建权限请求参数
 */
export interface PermissionCreate {
  name: string;
  code: string;
  resource: string;
  action: string;
  type?: PermissionType;
  description?: string;
  parent_id?: string;
  sort_order?: number;
  is_active?: boolean;
}

/**
 * 更新权限请求参数
 */
export interface PermissionUpdate {
  name?: string;
  description?: string;
  parent_id?: string;
  sort_order?: number;
  is_active?: boolean;
}

/**
 * 权限列表响应
 */
export interface PermissionsResponse {
  items: Permission[];
  total: number;
  skip: number;
  limit: number;
}

/**
 * 角色信息
 */
export interface Role {
  id: string;
  name: string;
  code: string;
  team_code?: string; // 团队代码（undefined/null表示全局角色）
  description?: string;
  is_active: boolean;
  permissions: Permission[];
  created_at: string;
  updated_at: string;
}

/**
 * 创建角色请求参数
 */
export interface RoleCreate {
  name: string;
  code: string;
  description?: string;
  is_active?: boolean;
  permission_ids?: string[];
}

/**
 * 更新角色请求参数
 */
export interface RoleUpdate {
  name?: string;
  description?: string;
  is_active?: boolean;
  permission_ids?: string[];
}

/**
 * 角色列表响应
 */
export interface RolesResponse {
  items: Role[];
  total: number;
  skip: number;
  limit: number;
}

/** 权限分组接口返回：平铺列表 + 展示顺序/文案，前端用 groupBy 按类型组合 */
export interface PermissionsGroupedRawResponse {
  items: Permission[];
  resource_order?: string[];
  resource_labels?: Record<string, string>;
}

/** 前端用 groupBy 从平铺数据组合后的结构，供权限管理/角色分配使用 */
export interface PermissionsGroupedResponse {
  menu: {
    route: Record<string, Permission[]>;
    button: Record<string, Permission[]>;
  };
  api: Record<string, Permission[]>;
  resource_order?: string[];
  resource_labels?: Record<string, string>;
}

/**
 * 获取权限平铺列表及展示顺序/文案（用于权限管理/角色分配；前端用 groupBy 按类型组合）
 */
export async function getPermissionsGrouped(params?: {
  is_active?: boolean;
}): Promise<PermissionsGroupedRawResponse> {
  const queryParams = new URLSearchParams();
  if (params?.is_active !== undefined) queryParams.append("is_active", params.is_active.toString());
  const q = queryParams.toString();
  return apiRequest<PermissionsGroupedRawResponse>(
    `/admin/rbac/permissions/grouped${q ? `?${q}` : ""}`
  );
}

/**
 * 获取权限列表
 */
export async function getPermissions(params?: {
  skip?: number;
  limit?: number;
  resource?: string;
  is_active?: boolean;
  type?: "menu" | "api";
}): Promise<PermissionsResponse> {
  const queryParams = new URLSearchParams();
  if (params?.skip !== undefined) queryParams.append("skip", params.skip.toString());
  if (params?.limit !== undefined) queryParams.append("limit", params.limit.toString());
  if (params?.resource !== undefined) queryParams.append("resource", params.resource);
  if (params?.is_active !== undefined) queryParams.append("is_active", params.is_active.toString());
  if (params?.type !== undefined) queryParams.append("type", params.type);
  
  const queryString = queryParams.toString();
  const path = `/admin/rbac/permissions${queryString ? `?${queryString}` : ""}`;
  
  return apiRequest<PermissionsResponse>(path);
}

/**
 * 获取单个权限
 */
export async function getPermission(permissionId: string): Promise<Permission> {
  return apiRequest<Permission>(`/admin/rbac/permissions/${permissionId}`);
}

/**
 * 创建权限
 */
export async function createPermission(data: PermissionCreate): Promise<Permission> {
  return apiRequest<Permission>("/admin/rbac/permissions", {
    method: "POST",
    body: JSON.stringify(data),
  });
}

/**
 * 更新权限
 */
export async function updatePermission(permissionId: string, data: PermissionUpdate): Promise<Permission> {
  return apiRequest<Permission>(`/admin/rbac/permissions/${permissionId}`, {
    method: "PUT",
    body: JSON.stringify(data),
  });
}

/**
 * 更新菜单配置（支持团队级别的覆盖）
 */
export async function updateMenuConfig(
  permissionId: string,
  data: { parent_id?: string | null; sort_order: number }
): Promise<{
  id: string;
  permission_id: string;
  team_id: string | null;
  parent_id: string | null;
  sort_order: number;
}> {
  return apiRequest(`/admin/rbac/menus/${permissionId}/config`, {
    method: "PUT",
    body: JSON.stringify(data),
  });
}

/**
 * 删除权限
 */
export async function deletePermission(permissionId: string): Promise<void> {
  return apiRequest<void>(`/admin/rbac/permissions/${permissionId}`, {
    method: "DELETE",
  });
}

/**
 * 获取角色列表
 */
export async function getRoles(params?: {
  skip?: number;
  limit?: number;
  is_active?: boolean;
}): Promise<RolesResponse> {
  const queryParams = new URLSearchParams();
  if (params?.skip !== undefined) queryParams.append("skip", params.skip.toString());
  if (params?.limit !== undefined) queryParams.append("limit", params.limit.toString());
  if (params?.is_active !== undefined) queryParams.append("is_active", params.is_active.toString());
  
  const queryString = queryParams.toString();
  const path = `/admin/rbac/roles${queryString ? `?${queryString}` : ""}`;
  
  return apiRequest<RolesResponse>(path);
}

/**
 * 获取单个角色
 */
export async function getRole(roleId: string): Promise<Role> {
  return apiRequest<Role>(`/admin/rbac/roles/${roleId}`);
}

/**
 * 创建角色
 */
export async function createRole(data: RoleCreate): Promise<Role> {
  return apiRequest<Role>("/admin/rbac/roles", {
    method: "POST",
    body: JSON.stringify(data),
  });
}

/**
 * 更新角色
 */
export async function updateRole(roleId: string, data: RoleUpdate): Promise<Role> {
  return apiRequest<Role>(`/admin/rbac/roles/${roleId}`, {
    method: "PUT",
    body: JSON.stringify(data),
  });
}

/**
 * 删除角色
 */
export async function deleteRole(roleId: string): Promise<void> {
  return apiRequest<void>(`/admin/rbac/roles/${roleId}`, {
    method: "DELETE",
  });
}

/**
 * 菜单项（树形结构）
 */
export interface MenuItem {
  id: string;
  name: string;
  code: string;
  resource: string;
  action: string;
  type: PermissionType;
  description?: string;
  parent_id?: string;
  parentId?: string;  // camelCase 格式（与 parent_id 相同）
  sort_order?: number;
  sortOrder?: number;  // camelCase 格式（与 sort_order 相同）
  is_active: boolean;
  children?: MenuItem[];
}

/**
 * 获取菜单树
 */
export async function getMenuTree(): Promise<MenuItem[]> {
  return apiRequest<MenuItem[]>("/admin/rbac/menus/tree");
}
